pkg/upstream/write.go

```diff
@@ -198,3 +232,557 @@ func maybeEncryptIdentityConfig(identityConfig *kotsv1beta1.IdentityConfig) ([]b
 
 	return b.Bytes(), nil
 }
+
+// FindReplicatedSDKChart will look for a chart with the name "replicated" or "replicated-sdk" in the archive
+// and return the name of the chart and whether it is the parent chart or a subchart
+func FindReplicatedSDKChart(archive io.Reader, replicatedChartNames []string) (string, bool, error) {
+	replicatedChartNamesMap := map[string]bool{}
+	for _, chartName := range replicatedChartNames {
+		replicatedChartNamesMap[chartName] = true
+	}
+
+	gzReader, err := gzip.NewReader(archive)
+	if err != nil {
+		return "", false, errors.Wrap(err, "failed to create gzip reader")
+	}
+	defer gzReader.Close()
+
+	tarReader := tar.NewReader(gzReader)
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return "", false, errors.Wrap(err, "failed to read header from tar")
+		}
+
+		switch header.Typeflag {
+		case tar.TypeReg:
+			if filepath.Base(header.Name) == "Chart.yaml" {
+				content, err := io.ReadAll(tarReader)
+				if err != nil {
+					return "", false, errors.Wrapf(err, "failed to read file %s", header.Name)
+				}
+
+				chartInfo := struct {
+					ChartName string `json:"name" yaml:"name"`
+				}{}
+				if err := yaml.Unmarshal(content, &chartInfo); err != nil {
+					return "", false, errors.Wrapf(err, "failed to unmarshal %s", header.Name)
+				}
+
+				if replicatedChartNamesMap[chartInfo.ChartName] {
+					// check if the sdk is the parent chart or a subchart based on the path
+					replicatedSDKChartName := chartInfo.ChartName
+					isReplicatedSDK := !strings.Contains(filepath.Dir(header.Name), string(os.PathSeparator))
```

Comment:
wouldn't there be a top-level directory before the filename? e.g. `replicated/Chart.yaml`.